JAVA-2297: Set socket options from NettyOptions#afterBootstrapInitialized

olim7t opened the pull request:
Motivation:

Setting socket options on the bootstrap is currently hardcoded in
ChannelFactory. This can cause warnings if the transport is overridden
with an implementation that does not support some of these options (ex:
TCP_NODELAY for Unix domain sockets).

Modifications:

Move the code to NettyOptions#afterBootstrapInitialized.

Result:

Custom NettyOptions implementations can decide which socket options to
set. Note that they will now have to copy over the code from
DefaultNettyOptions to get the default behavior.

olim7t committed:
JAVA-2297: Set socket options from NettyOptions#afterBootstrapInitialized

Motivation:

Setting socket options on the bootstrap is currently hardcoded in
ChannelFactory. This can cause warnings if the transport is overridden
with an implementation that does not support some of these options (ex:
TCP_NODELAY for Unix domain sockets).

Modifications:

Move the code to NettyOptions#afterBootstrapInitialized.

Result:

Custom NettyOptions implementations can decide which socket options to
set. Note that they will now have to copy over the code from
DefaultNettyOptions to get the default behavior.

tolbertam commented on core/src/main/java/com/datastax/oss/driver/internal/core/context/NettyOptions.java:
```diff
@@ -62,6 +66,45 @@
+   *     set will typically be read from).
+   * @since 4.1.0
+   */
+  default void setSocketOptions(Bootstrap bootstrap, DriverExecutionProfile defaultProfile) {
```
One thing sort of awkward about this is that this method provides `DriverExecutionProfile` where the others don't. 

I was thinking maybe we could overload `afterBootstrapInitialized` with another method providing the profile as a parameter, but it gets sort of confusing to have two methods with the same name, and as you explained we don't want to break existing custom implementations which implicitly depended on the options to be set otherwise, so I think what you did here is probably our best option.

olim7t replied:
> One thing sort of awkward about this is that this method provides `DriverExecutionProfile` where the others don't.

Yes, that's the only way I found to provide a default implementation.

adutra replied:
Do we really expect people extending or implementing `NettyOptions`? This being an internal class, I would be in favor of being more agressive with backwards compatibility.

olim7t replied:
Very good point, for some reason I was assuming it was in the public API and didn't even check the package name.
You're right, it's fair to be a bit more aggressive here, I'll do the changes.

tolbertam reviewed (approved):
:+1:

emerkle826 commented on core/src/main/java/com/datastax/oss/driver/internal/core/context/NettyOptions.java:
```diff
@@ -62,6 +66,45 @@
+   */
+  default void setSocketOptions(Bootstrap bootstrap, DriverExecutionProfile defaultProfile) {
+    boolean tcpNoDelay = defaultProfile.getBoolean(DefaultDriverOption.SOCKET_TCP_NODELAY);
+    bootstrap.option(ChannelOption.TCP_NODELAY, tcpNoDelay);
```
Do we want to use the same `isDefined()` check here for the Unix socket case, or is the idea that when using Unix sockets, you would provide a different implementation for `setSocketOptions` that does not try to set TCP_NODELAY?

emerkle826 replied:
I commented too soon. We are doing it this way to keep the functionality backward compatible, correct?

emerkle826 reviewed (approved):
Ignore my comments, I overthought the change too much. :+1:

olim7t commented:
Force-pushed to do it in `afterBootstrapInitialized`. I added a note in the upgrade guide since this is still a relatively high-level feature.

adutra commented on core/src/main/java/com/datastax/oss/driver/internal/core/context/DefaultNettyOptions.java:
```diff
@@ -131,7 +134,30 @@ public ByteBufAllocator allocator() {
+    }
   }
 
   @Override
```
Not related to this PR, but I noticed that `PromiseCombiner` has been deprecated (used in method `onClose` below). Is there anything we can do to avoid using it?

olim7t replied:
It's just the no-arg constructor that's deprecated. I guess we should inline its implementation: `new PromiseCombiner(ImmediateEventExecutor.INSTANCE)`; we can't use any of the driver's executors since they are being shut down.
I'll add the change in a separate commit.

olim7t replied:
Actually `GlobalEventExecutor.INSTANCE` is a better choice (see the explanations on netty/netty@625c4e8286d3c8a36046feb1289b6b710c7541a5).

olim7t committed:
Revisit NettyOptions#onClose

`new PromiseCombiner()` is deprecated (see netty/netty@625c4e8).
PromiseCombiner is not thread-safe, all of its methods must be invoked
from the same event loop.

We can't use any of the driver's event loops since we're shutting them
down, so use GlobalEventExecutor.INSTANCE.

To respect the contract of PromiseCombiner, schedule its construction on
that event loop instead of doing it from the current thread.
